clippy_lints/src/thread_local_initializer_can_be_made_const.rs

```diff
@@ -57,6 +57,48 @@ impl ThreadLocalInitializerCanBeMadeConst {
 
 impl_lint_pass!(ThreadLocalInitializerCanBeMadeConst => [THREAD_LOCAL_INITIALIZER_CAN_BE_MADE_CONST]);
 
+#[inline]
+fn is_thread_local_initializer(
+    cx: &LateContext<'_>,
+    fn_kind: rustc_hir::intravisit::FnKind<'_>,
+    span: rustc_span::Span,
+) -> bool {
+    if in_external_macro(cx.sess(), span)
+        && let Some(callee) = span.source_callee()
+        && let Some(macro_def_id) = callee.macro_def_id
+        && cx.tcx.is_diagnostic_item(thread_local_macro, macro_def_id)
+        && let intravisit::FnKind::ItemFn(..) = fn_kind
+    {
+        return true;
+    }
+    false
+}
+
+#[inline]
+fn initializer_fn_is_const(cx: &LateContext<'_>, defid: rustc_span::def_id::DefId) -> bool {
+    // check that the initializer function is not already `const`.
+    // this done for `os_local` fallback as that implementation includes an `__init()`
+    // function regardless of whether the `const` keyword is used.
+    // For const-qualified `thread_local!`s, the `__init` function is a const fn.
+    // For non-const-qualified `thread_local!`s, the `__init` function is not a const fn.
+    // we are going to check that we are in a non-const-qualified `thread_local!` and
+    // then that the `__init` contents are const-qualifiable. If they are,
+    // we can suggest to make the `__init` function const.
```

Comment:
Honestly, I'm not sure if these comments are necessary. I tried reading them, and they are not really clear. So I was thinking of making it smaller and linking to this PR, but even then, it's pretty clear what the function does, what the result is and knowing the underlying reason isn't really necessary IMO.

But, what do you think?

Reply:
I think we can just keep the call to `txc` and move it back with the main if, what do you think?

Reply:
I'd say let's remove it, `is_thread_local_initializer` followed by `!tcx.is_const_fn` for the same defid is fairly obvious I think.